### ISSUE TITLE
Predict on GraphConvTensorGraph returns a vector with the wrong length (batch padding?)

### ISSUE THREAD

**Issue by miguelroboso** [deepchem/deepchem#841]
I tried to run deepchem on my data, and after training, I am trying to check which of the inputs have been predicted correctly, and which ones are not.

```
predicted_val = model.predict(valid_dataset)
true_val = valid_dataset.y
print(predicted_val.shape)
print(true_val.shape)
```

The issue is that `predicted_val.shape` is different from `true_val.shape`, and that should not happen. 

I tried to use PyCharm to check what `predict` was called:

It seems to be the `tensor_graph.py` predict function, 

which sets `pad_batches=False` as expected, 

and calls `default_generator` of  `GraphConvTensorGraph `.

**Comment by rbharath** [comment]
@miguelroboso Thanks for the error report. Would you be able to create a minimal failing example that's standalone? (Perhaps using a MoleculeNet dataset for convenience). It would help to be able to reproduce the bug locally for debugging.

**Comment by miguelroboso** [comment]
```

# coding: utf-8

# ### Basic imports.


from __future__ import division
from __future__ import print_function
from __future__ import unicode_literals

import numpy as np
import tensorflow as tf
import deepchem as dc
from deepchem.models.tensorgraph.models.graph_models import GraphConvTensorGraph


BATCH_SIZE = 128

# ### Load dataset

#
# get_ipython().magic(u'load_ext autoreload')
# get_ipython().magic(u'autoreload 2')
# get_ipython().magic(u'pdb off')

from deepchem.utils.save import load_from_disk

dataset_file= "testInput.csv"
dataset = load_from_disk(dataset_file)
print("Columns of dataset: %s" % str(dataset.columns.values))
print("Number of examples in dataset: %s" % str(dataset.shape[0]))

import deepchem as dc

# featurizer = dc.feat.CircularFingerprint(size=1024)
featurizer = dc.feat.ConvMolFeaturizer()

loader = dc.data.CSVLoader(
      tasks=["active"], smiles_field="SMILES",
      featurizer=featurizer)
dataset = loader.featurize(dataset_file)

splitter = dc.splits.ScaffoldSplitter(dataset_file)
train_dataset, valid_dataset, test_dataset = splitter.train_valid_test_split(
    dataset, seed = 1138)

transformers = [dc.trans.NormalizationTransformer(transform_y=True, dataset=train_dataset)]

for dataset in [train_dataset, valid_dataset, test_dataset]:
  for transformer in transformers:
      dataset = transformer.transform(dataset)

model = GraphConvTensorGraph(
    1, batch_size=128, mode='classification')
model.fit(train_dataset, nb_epoch=10)


# Let's try to evaluate the performance of the model we've trained. For this, we need to define a metric, a measure of model performance. `dc.metrics` holds a collection of metrics already. For this dataset, it is standard to use the ROC-AUC score, the area under the receiver operating characteristic curve (which measures the tradeoff between precision and recall). Luckily, the ROC-AUC score is already available in DeepChem.
#
# To measure the performance of the model under this metric, we can use the convenience function `model.evaluate()`.


metric = dc.metrics.Metric(
    dc.metrics.roc_auc_score, np.mean, mode="classification")

print("Evaluating model")
train_scores = model.evaluate(train_dataset, [metric], transformers)
print("Training ROC-AUC Score: %f" % train_scores["mean-roc_auc_score"])
valid_scores = model.evaluate(valid_dataset, [metric], transformers)
print("Validation ROC-AUC Score: %f" % valid_scores["mean-roc_auc_score"])

predicted_val = model.predict(valid_dataset)
true_val = valid_dataset.y

print(predicted_val.shape)

print(true_val.shape)
```

**Comment by miguelroboso** [comment]
This should run. I added a file with 5000 inputs. Validation is then 500. Predicted shape is (512,2)

**Comment by miaecle** [comment]
@rbharath I think the problem is from GraphGather layer, which use a tf.dynamic_partition. 
As in ```deepchem/deepchem/models/tensorgraph/layers.py```:
```python
class GraphGather(Layer):

  def __init__(self, batch_size, activation_fn=None, **kwargs):
    self.batch_size = batch_size
    self.activation_fn = activation_fn
    super(GraphGather, self).__init__(**kwargs)

  def create_tensor(self, in_layers=None, set_tensors=True, **kwargs):
    inputs = self._get_input_tensors(in_layers)

    # x = [atom_features, deg_slice, membership, deg_adj_list placeholders...]
    atom_features = inputs[0]

    # Extract graph topology
    membership = inputs[2]

    # Perform the mol gather

    assert self.batch_size > 1, "graph_gather requires batches larger than 1"

    # Obtain the partitions for each of the molecules
    activated_par = tf.dynamic_partition(atom_features, membership,
                                         self.batch_size)

    # Sum over atoms for each molecule
    sparse_reps = [
        tf.reduce_mean(activated, 0, keep_dims=True)
        for activated in activated_par
    ]
    max_reps = [
        tf.reduce_max(activated, 0, keep_dims=True)
        for activated in activated_par
    ]

    # Get the final sparse representations
    sparse_reps = tf.concat(axis=0, values=sparse_reps)
    max_reps = tf.concat(axis=0, values=max_reps)
    mol_features = tf.concat(axis=1, values=[sparse_reps, max_reps])

    if self.activation_fn is not None:
      mol_features = self.activation_fn(mol_features)
    out_tensor = mol_features
    if set_tensors:
      self.out_tensor = out_tensor
    return out_tensor
```
This layer constantly outputs `batch_size` in the first dimension(which maybe less than that in case of `pad_batches=False`).

**Comment by miguelroboso** [comment]
I was thinking that, as a temporary workaround, I could subset the split datasets as a multiple of my chosen batch size.
